snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,517 @@
+	// AWSElasticBlockStore represents an AWS Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
+	// +optional
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSnapshotSource `json:"awsElasticBlockStore,omitempty"`
+
+    // ... etc. for other snapshot types
+}
+```
+
+An example of the `VolumeSnapshotDataSource` for Amazon EBS snapshots:
+
+```
+// AWS EBS volume snapshot source
+type AWSElasticBlockStoreVolumeSnapshotSource struct {
+	// Unique id of the persistent disk snapshot resource. Used to identify the disk snapshot in AWS
+	SnapshotID string `json:"snapshotId"`
+}
+```
+
+### Event loop in controller
+
+The volume snapshot controller maintains two data structures (ActualStateOfWorld and DesiredStateOfWorld) and
+periodically reconciles the two. The data structures are being update by the API sever event handlers.
+
+* If a new `VolumeSnapshot` is added, the *add* handler adds it to the DesiredStateOfWorld (DSW)
+
+* If a `VolumeSnapshot` is deleted, the *delete* handler removes it from the DSW.
+
+* Reconciliation loop in the controller
+
+    * For every `VolumeSnapshot` in the ActualStateOfWorld (ASW) find the corresponding `VolumeSnapshot` in the DSW.
+    If such a snapshot does not exist, start a snapshot deletion operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a delete operation: only one such operation is allowed to exist for the given `VolumeSnapshot` and
+        `VolumeSnapshotData` pair
+
+        * The operation is an asynchronous function using the volume plugin to delete the actual snapshot in the back-end.
+
+        * When the plugin finishes deleting the snapshot, delete the `VolumeSnapshotData` referencing it and remove the
+        `VolumeSnapshot` reference from the ASW
+
+    * For every `VolumeSnapshot` in the DSW find the corresponding `VolumeSnapshot` in the ASW. If such a snapshot
+    does not exist, start a snapshot creation operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a volume snapshot creation operation: only one such operation is allowed to exist for the given
+        `VolumeSnapshot` and `VolumeSnapshotData` pair.
+
+        * The operation is an asynchronous function using the volume plugin to create the actual snapshot in the back-end.
+
+        * When the plugin finishes creating the snapshot a new `VolumeSnapshotData` is created holding a reference to
+        the actual volume snapshot.
+
+    * For every snapshot present in the ASW and DSW find its `VolumeSnapshotData` and verify the bi-directional
+    binding is correct: if not, update the `VolumeSnapshotData` reference.
```

Comment:
The binding is incorrect when it's incomplete. The code should assume the API writes are unreliable and may fail. When the user creates the VolumeSnapshot its VolumeSnapshotData reference is empty, then the controller creates the snapshot and VolumeSnapshotData: the VolumeSnapshotData is always created only if the actual snapshot creation succeeded, and always with reference to VolumeSnapshot. Then the VolumeSnapshot object is updated (this may fail). So the VolumeSnapshot may exits with missing reference to the VolumeSnapshotData and this has to be fixed...

Thinking about it: I can leave this part out eventually since it sounds like an implementation detail (there are more ways to ensure the objects consistency).